test/fail_compilation/fail14596.d

```diff
@@ -0,0 +1,10 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/fail14596.d(9): Error: cannot cast expression malloc(42LU) of type void* to type char[]
+---
+*/
+void main() {
+    import core.stdc.stdlib : malloc;
+    auto ptr = cast(char[]) malloc(42);
+}
```

Comment:
As a better place, you can add the new test case to `fail_compilation/fail_casting.d`.